### ISSUE TITLE
Parallelization

### ISSUE THREAD

**Issue by alexW432** [leeping/forcebalance#83]
Hi,
I am running a ForceBalance optimization with 369 tip4p water molecules with a custom non-bonded force and:
liquid_eq_steps 50000
liquid_md_steps 500000
gas_eq_steps 50000
gas_md_steps 500000
liquid_timestep 2.0
liquid_interval 0.2
However it takes around 24 hours to finish a step. Does force balance support parallelization to help reduce this time? If not is there anything else I could do to speed this up?
Many Thanks,
Alex

**Comment by leeping** [comment]
Hi Alex,

Yes, there are some options for parallelization.  You should check npt.out to see if OpenMM is using an optimized platform (i.e. CUDA if you have an NVidia GPU).  I'm not sure what platform it's currently using, given the speeds you're seeing.

You can also parallelize over T,P phase points using the Work Queue interface (e.g. if you are fitting the temperature dependence of properties).  

Basically, you first run ForceBalance on a machine we call `master_host` and add `wq_port 5678` in $options.  

Then you install ForceBalance and the MD software on a remote machine, and run `work_queue_worker master_host 5678`. 

The worker will connect to the master host, which assigns a MD run to it, so you can have workers runnig multiple simulations in parallel.

Thanks,

- Lee-Ping

**Comment by alexW432** [comment]
Hi Lee-Ping, 
I don't have a Nividia GPU and I think this may be the problem. I specify in the file.in 'openmm_platform Reference' however if I chech the npt.out it reads 
```Python
coords                    liquid.pdb
mol                       <forcebalance.molecule.Molecule object at 0x3046f90>
pbc                       True
platname                  CUDA
FF                        <forcebalance.forcefield.FF object at 0x3046810>
```
and 'Platform CUDA does not exist (valid options are ['Reference', 'CPU'] (case-sensitive))'. should I be asking to use CPU or Reference in a different way?
Thanks,
Alex

**Comment by leeping** [comment]
Hi Alex,

I think custom forces also work on the CPU or OpenCL platforms.  If you have an AMD GPU, you might want to try the OpenCL platform.  @peastman: Can you confirm that custom forces are implemented on all four platforms?

Most of my OpenMM calculations run on boxes with an NVidia GPU (the gaming cards that cost ~$500 are fine).

Thanks,

- Lee-Ping

**Comment by alexW432** [comment]
I see is it normal for me to specify
```Python
openmm_platform CPU
```
however still have npt.out say
```Python
platname                  CUDA
```
Thanks 
Alex

**Comment by alexW432** [comment]
This is npt.out, I am confused since in the options it says platname CPU as expected however later says platname CUDA 
```Python
#=======================#
# STARTING CALCULATION! #
#=======================#

python npt.py openmm 298.150 1.000
Work Queue library import fail (You can't queue up jobs using Work Queue)
^M#==================================================================#
#| ^[[1;94m ForceBalance condensed phase simulation using engine: OPENMM ^[[0m |#
#==================================================================#
^M#========================================================#
#| ^[[92m             Options from ForceBalance              ^[[0m |#
#========================================================#
force_cuda                False
mts_integrator            0
rpmd_beads                0
precision                 mixed
platname                  CPU
liquid_coords             liquid.pdb
gas_coords                gas.pdb
root                      /scratch/adw62/fbcb/forcebalance-616c5169809540bb3ca19c7823cecdee87cbd194/studies/4P3C
name                      WATER
type                      LIQUID_OPENMM
weight                    1.0
fdgrad                    0
fdhess                    0
fdhessdiag                0
sleepy                    0
fd1_pids                  []
fd2_pids                  []
h                         0.001
backup                    1
zerograd                  -1
epsgrad                   0.0
pgrad                     [0, 1, 2]
tgtdir                    targets/WATER
w_rho                     1.0
w_hvap                    1.0
w_alpha                   1.0
w_kappa                   1.0
w_cp                      1.0
w_eps0                    1.0
w_normalize               0
manual                    0
hvap_subaverage           0
liquid_eq_steps           50000
liquid_md_steps           500000
gas_eq_steps              50000
gas_md_steps              500000
liquid_timestep           2.0
liquid_interval           0.2
gas_timestep              1.0
gas_interval              0.1
adapt_errors              0
minimize_energy           1
self_pol_mu0              1.855
self_pol_alpha            1.47
anisotropic_box           0
save_traj                 0
n_molecules               369
rho_denom                 2.0
hvap_denom                5.0
alpha_denom               1.0
kappa_denom               10.0
cp_denom                  2.0
eps0_denom                10.0
----------------------------------------------------------
For the condensed phase system, I will collect 5000 snapshots spaced apart by 100 x 2.000 fs time steps
For the gas phase system, I will collect 5000 snapshots spaced apart by 100 x 1.000 fs time steps
There are 369 molecules in the liquid
^M#========================================================#
#| ^[[92m             Engine options for liquid              ^[[0m |#
#========================================================#
coords                    liquid.pdb
mol                       <forcebalance.molecule.Molecule object at 0x3851f90>
pbc                       True
platname                  CUDA
FF                        <forcebalance.forcefield.FF object at 0x3851810>
----------------------------------------------------------
^M#========================================================#
#| ^[[92m               Engine options for gas               ^[[0m |#
#========================================================#
coords                    gas.pdb
mol                       <forcebalance.molecule.Molecule object at 0x24e8f10>
pbc                       False
platname                  Reference
FF                        <forcebalance.forcefield.FF object at 0x3851810>
----------------------------------------------------------
Running without a target, using current directory.
Platform CUDA does not exist (valid options are ['Reference', 'CPU'] (case-sensitive))
^M#========================================================#
#| ^[[1;94m         Condensed phase molecular dynamics         ^[[0m |#
#========================================================#
```
Thank you in advance,
Alex

**Comment by leeping** [comment]
Hi Alex,

I just fixed the code in `npt.py` to take the `platname` argument from the target options.  Since I usually use the CUDA platform, I didn't catch this issue.  Can you try the latest commit and let me know if it works?

Thanks,

- Lee-Ping

**Comment by alexW432** [comment]
Hi Lee-Ping,

This seems to fix the issue it now read CPU in both 'options' and 'Engine options for liquid'. 
Thank you very much.
Alex

**Comment by peastman** [comment]
Yes, custom forces work on all platforms.
